src/dict.c

```diff
@@ -74,6 +74,17 @@ struct dictEntry {
     struct dictEntry *next; /* Next entry in the same hash bucket. */
 };
 
+typedef struct {
+    union {
+        void *val;
+        uint64_t u64;
+        int64_t s64;
+        double d;
+    } v;
+    struct dictEntry *next; /* Next entry in the same hash bucket. */
+    unsigned char data[];
+} embeddedDictEntry;
+
```

Comment:
[Readability suggestion] I think we can get rid of the magic constant, `ENTRY_METADATA_BYTES`. Also I think `key_buf` conveys the purpose of this memory region clearer than `data`

```suggestion
typedef struct {
    union {
        void *val;
        uint64_t u64;
        int64_t s64;
        double d;
    } v;
    struct dictEntry *next; /* Next entry in the same hash bucket. */
    unsigned char key_offset; /* offset into key_buf where the key is located at. */
    char key_buf[];
} embeddedDictEntry;
```

Reply:
I was going to ask, what is ENTRY_METADATA_BYTES even for?

Reply:
Just a note, the `unsigned char key_offset` will get padded to 8 bytes in a sizeof calculation, so we need to subtract off 7 when we do the math during the zmalloc otherwise we will over allocate memory. Either that or we need to allocate a normal dictEntry and just cast it to an embeddedDictEntry.

It also feels like we are tightly coupling the naming to the sds, which has a concept of having the start of the the key part way into the buffer with a header in the prefix. I would rather just call it key_header_size.

Reply:
> the unsigned char key_offset will get padded to 8 bytes

Good point. We could pack the struct ourselves and `static_assert` on `v` and `next` being aligned

```c
typedef struct __attribute__((packed)) {
    union {
        void *val;
        uint64_t u64;
        int64_t s64;
        double d;
    } v;
    struct dictEntry *next; /* Next entry in the same hash bucket. */
    unsigned char key_offset; /* offset into key_buf where the key is located at. */
    char key_buf[];
} embeddedDictEntry;
```

> I would rather just call it key_header_size.

SGTM

Reply:
Vitaly and I had also planned to abstract lot of these metadata values out to individual variables for better readability.

Reply:
Getting the following error with the packed struct. checking how to avoid it.

```
dict.c:930:37: error: taking address of packed member of ‘struct <anonymous>’ may result in an unaligned pointer value [-Werror=address-of-packed-member]
  930 |     if (entryIsEmbedded(de)) return &decodeEmbeddedEntry(de)->next;
      |                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors
```

Reply:
@hpatro I think you can do this? We only care about the last data being packed, the rest of the data will be properly aligned.
```
typedef struct {
    union {
        void *val;
        uint64_t u64;
        int64_t s64;
        double d;
    } v;
    struct dictEntry *next; /* Next entry in the same hash bucket. */
    typedef struct __attribute__((packed)) {
        unsigned char key_offset; /* offset into key_buf where the key is located at. */
        char key_buf[];
    }
} embeddedDictEntry;
```

Reply:
@madolson `sizeof` reported is still 24 bytes with the above. So, proceeding with your first suggestion. If anyone else has a better idea, will revisit this. Introduced a [compactSizeEmbeddedDictEntry](https://github.com/valkey-io/valkey/pull/541/commits/029917f56ae71a822d1f10889544936bc530dad6) method to find the size for allocation.